store/postgres/migrations/2019-03-11-221621_delete_unneeded_entity_history/up.sql

```diff
@@ -0,0 +1,190 @@
+-- Remove mention of data_after from the log_entity_event function
+CREATE OR REPLACE FUNCTION log_entity_event()
+    RETURNS trigger AS
+$$
+DECLARE
+    event_id INTEGER;
+    new_event_id INTEGER;
+    is_reversion BOOLEAN := FALSE;
+    operation_type INTEGER := 10;
+    event_source  VARCHAR;
+    subgraph VARCHAR;
+    entity VARCHAR;
+    entity_id VARCHAR;
+    data_before JSONB;
+BEGIN
+    -- Get operation type and source
+    IF (TG_OP = 'INSERT') THEN
+        operation_type := 0;
+        event_source := NEW.event_source;
+        subgraph := NEW.subgraph;
+        entity := NEW.entity;
+        entity_id := NEW.id;
+        data_before := NULL;
+    ELSIF (TG_OP = 'UPDATE') THEN
+        operation_type := 1;
+        event_source := NEW.event_source;
+        subgraph := OLD.subgraph;
+        entity := OLD.entity;
+        entity_id := OLD.id;
+        data_before := OLD.data;
+    ELSIF (TG_OP = 'DELETE') THEN
+        operation_type := 2;
+        event_source := current_setting('vars.current_event_source', TRUE);
+        subgraph := OLD.subgraph;
+        entity := OLD.entity;
+        entity_id := OLD.id;
+        data_before := OLD.data;
+    ELSE
+        RAISE EXCEPTION 'unexpected entity row operation type, %', TG_OP;
+    END IF;
+
+    IF event_source = 'REVERSION' THEN
+        is_reversion := TRUE;
+    END IF;
+
+    SELECT id INTO event_id
+    FROM event_meta_data
+    WHERE db_transaction_id = txid_current();
+
+    new_event_id := null;
+
+    IF event_id IS NULL THEN
+        -- Log information on the postgres transaction for later use in revert operations
+        INSERT INTO event_meta_data
+            (db_transaction_id, db_transaction_time, source)
+        VALUES
+            (txid_current(), statement_timestamp(), event_source)
+        RETURNING event_meta_data.id INTO new_event_id;
+    END IF;
+
+    -- Log row metadata and changes, specify whether event was an original ethereum event or a reversion
+    INSERT INTO entity_history
+        (event_id, entity_id, subgraph, entity, data_before, reversion, op_id)
+    VALUES
+        (COALESCE(new_event_id, event_id), entity_id, subgraph, entity, data_before, is_reversion, operation_type);
+    RETURN NULL;
+END;
+$$ LANGUAGE plpgsql;
+
+-- Remove unneeded entity_history data, and event_meta_data that becomes
+-- unnecessary because of that, too. Since we need to get rid of most of
+-- the data, it is faster to copy the 'good' data into a new table and swap
```

Comment:
Curious: Is it worth doing this copy, recreate indexes, drop table, rename dance just for performance? I prefer simplicity over performance (until performance becomes a problem).

Reply:
The simple version of all this is
```
delete from entity_history where subgraph = 'subgraphs' or data_before = data_after;
vacuum (full, analyze) entity_history;
delete from event_meta_data d where not exists (select 1 from entity_history h where h.event_id = d.id);
vacuum (full, analyze) event_meta_data;
```

but that takes at least twice as long as this script (and the more data we accumulate that needs to be deleted, the bigger the advantage of this script), and the vacuum will also lock the corresponding tables, effectively preventing any writes to `entities` (because of the `log_entity_event` trigger).

If you want more precise timing numbers, I could do that on a clone of the prod db.